deck-of-cards-kata/src/main/java/bnymellon/codekatas/deckofcards/custom/collections/ImmutableHashBag.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2020 The Bank of New York Mellon.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package bnymellon.codekatas.deckofcards.custom.collections;
+
+import java.util.AbstractCollection;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+final class ImmutableHashBag<T>
+        extends AbstractCollection<T>
+        implements ImmutableBag<T>
+{
+    private final MutableBag<T> hashBag;
+
+    public ImmutableHashBag(Collection<T> collection)
+    {
+        HashBag<T> bag = new HashBag<>();
```

Comment:
Effective Java principles suggest to have interface on the left.

Reply:
It might be a good to have another constructor on HashBag to make the code succinct.

Reply:
Makes sense. I'll make the change.